File: 2010_2011_mf_sujet_1_introduction_matlab/TP1_MF.py
"""TP1 Master Maths Fondamentales — Introduction à Python (NumPy/Matplotlib)."""
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
from scipy.integrate import quad
from scipy.optimize import fsolve

from toto import toto
from q12 import q12

# Variables
a = 4; b = np.pi
print(f'a={a}, b={b}')
del b
print(f'a={a}')
print(np.sqrt(a))

A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(A); print(A[1, 2])

X = np.arange(1, 21); print(X)
Y = np.arange(1, 21, 2); print(Y)
print(np.arange(1, -2.5, -0.5))

b_row = A[1, :]; c_col = A[:, 2]; B_sub = A[0:2, 0:2]; C_sub = A[:, 0:2]

print(A.shape[0])
C = A.T
print(A + C); print(A @ C)

n = 4
I4 = np.eye(n); B = np.zeros((n, n)); D = np.ones((n, n))
print(np.linalg.matrix_power(A, 2))
print(np.linalg.det(A), np.trace(A), np.linalg.eig(A)[0])

E = 5 * np.ones((3, 3))
print(A[:, 0] @ A[:, 1])   # A(:,1)'*A(:,2)
print(A > 2)
rca = np.sqrt(np.arange(2, 12, 2))
print(A**2); print(A * C)

V = np.arange(1, 11)**2
W = 1 / np.arange(1, 11)

# Hilbert matrix
M = np.zeros((3, 3))
for i in range(3):
    for j in range(3):
        M[i, j] = 1 / (i + j + 1)
print(M); print(np.linalg.det(M)); print(np.linalg.inv(M)); print(np.linalg.eig(M)[0])

# Plots
t = np.arange(0, 3.1, 0.1)
plt.figure(); plt.plot(t, np.exp(t) + 3); plt.show()
plt.figure(); plt.plot(t, np.exp(t), t, 1 + t); plt.show()

# %% 11
plt.figure()
plt.plot(t, np.cos(t), t, np.sin(t), t, t**2)
plt.show()

x_pi = np.linspace(-np.pi, np.pi, 200)
plt.figure()
plt.plot(x_pi, np.cos(x_pi)); plt.plot(x_pi, np.sin(x_pi)); plt.plot(x_pi, x_pi**2)
plt.show()

# Functions
print(toto(2))
x01 = np.linspace(-1, 1, 200)
plt.figure(); plt.plot(x01, toto(x01)); plt.show()

result, _ = quad(toto, 0, 1); print(result)
root = fsolve(toto, 0)[0]; print(root)

# %% 12
plt.figure(); plt.plot(np.linspace(0, 1, 200), q12(np.linspace(0, 1, 200))); plt.show()
result12, _ = quad(q12, 0, 1); print(result12)
print(np.pi / 2 - 1)

toto2 = lambda x: np.cos(x)**2 - np.sin(x)**2
result2, _ = quad(toto2, 0, np.pi); print(result2)
root2 = fsolve(lambda x: x**3 - 3, 0)[0]; print(root2)

# %% 13
root3 = fsolve(lambda x: np.exp(x) - 3*x, 0)[0]; print(root3)

# %% Statistics
X = np.random.rand(1000)
s = np.sum(X); m = np.mean(X); v = np.var(X, ddof=1); sigma = np.std(X, ddof=1)
print(f's={s}, m={m}, v={v}, sigma={sigma}')

# %% 14
print(np.mean(X**3))

# %% Histograms
X = np.random.rand(100)
plt.figure(); plt.hist(X); plt.show()

X = np.random.rand(1000)
plt.figure(); plt.hist(X); plt.show()

X = np.random.rand(10000)
plt.figure(); plt.hist(X, bins=100, density=True); plt.show()
plt.figure(); plt.hist(X, bins=20, density=True); plt.show()

# %% 16 — Normal
X = np.random.randn(1000)
plt.figure()
plt.hist(X, bins=20, density=True)
x = np.linspace(-3, 3, 200)
plt.plot(x, norm.pdf(x), 'r')
plt.show()
print(np.mean(X), np.std(X, ddof=1))

File: 2010_2011_mf_sujet_1_introduction_matlab/TP1_MF.m
a=4;b=pi
who
clear b
who
sqrt(a)
A=[1 2 3;4 5 6;7 8 9]
A(2,3)
%% Utilisation de :
X=[1:20]
Y=[1:2:20]
[1:-0.5:-2]
%% 2
1+1:5
1+(1:5)
%% 
b=A(2,:)
c=A(:,3)
B=A(1:2,1:2)
%% 3
C=A(:,1:2)
%% Operations sur les matrices
help elmat
length(A)
C=A'
A+C
A*C
%%
n=4;I4=eye(n)
B=zeros(n)
D=ones(n)
p=2;A^p
det(A),trace(A),eig(A)
ans
%% 4
E=5*ones(3)
%% 5
A(:,1)'*A(:,2)
%%
A>2
%% 6
rca=sqrt(2:2:10)
%% Utilisation de .
A^2
A.^2
%%
A*C
A.*C
%% 7-8
V=[1:10].^2
W=1./[1:10]
%% Boucles - 9
M=zeros(3);
for i =1:3
    for j=1:3
        M(i,j)=1/(i+j-1);
    end
end
M
%%
det(M)
inv(M)
eig(M)
%% Representations graphiques
t=0:1/10:3; plot(t,exp(t)+3)
%%
t=0:1/10:3; plot(t,exp(t),t,1+t)
%% 11
plot(t,cos(t),t,sin(t),t,t.^2)
% ou bien
%% 11
fplot('cos',[-pi pi])
hold on
fplot('sin',[-pi pi])
t=[-pi:0.1:pi]; plot(t,t.^2)
hold off
%% Fonctions
toto(2)
fplot('toto',[-1 1])
%%
fplot(@toto,[-1 1])
%%
quad(@toto,0,1)
%%
fzero(@toto,0)
%% 12
fplot(@q12,[0 1])
%% 12
quad(@q12,0,1)
%% 12
pi/2-1
%%
toto2=@(x)cos(x).^2-sin(x).^2;
quad(toto2,0,pi)
fzero(@(x)x.^3-3,0)
%% 13
fzero(@(x)exp(x)-3*x,0)
%% Statistiques
rand
randn
X=rand(1000,1);
s=sum(X)
m=mean(X)
v=var(X)
sigma=std(X)
%% 14
mean(X.^3)
%Z=(X-m)/sigma;
%mean(Z),std(Z)
%% Histogrammes 15
help hist
X=rand(100,1);
histo(X)
%%
X=rand(1000,1);
figure
histo(X)
%%
X=rand(10000,1);
figure
histo(X,100,0,1)
figure
histo(X,20,0,1)
%%
X=rand(100000,1);
figure
histo(X,100,0,1)
hold on
histo(X,1000,1,1)
hold off

%% 16
X=randn(1000,1);
histo(X,20,0,1)
hold on
fplot(@(x)1./sqrt(2*pi)*exp(-x.^2./2),[-3 3],'r')
% ou bien
fplot('dnorm',[-3 3],'g')
hold off
mean(X)
std(X)